packages/phone-number-privacy/combiner/src/pnp/services/account-services.ts

```diff
@@ -0,0 +1,90 @@
+import { ContractKit } from '@celo/contractkit'
+import {
+  ErrorMessage,
+  FULL_NODE_TIMEOUT_IN_MS,
+  getDataEncryptionKey,
+  RETRY_COUNT,
+  RETRY_DELAY_IN_MS,
+} from '@celo/phone-number-privacy-common'
+import Logger from 'bunyan'
+import { LRUCache } from 'lru-cache'
+import { OdisError, wrapError } from '../../common/error'
+import { traceAsyncFunction } from '../../common/tracing-utils'
+
+export interface AccountService {
+  getAccount(address: string): Promise<string>
+}
+
+export interface ContractKitAccountServiceOptions {
+  fullNodeTimeoutMs: number
+  fullNodeRetryCount: number
+  fullNodeRetryDelayMs: number
+}
+
+export class CachingAccountService implements AccountService {
+  private cache: LRUCache<string, string, any>
+  constructor(baseService: AccountService) {
+    this.cache = new LRUCache({
+      max: 500,
+      ttl: 5 * 1000, // 5 seconds
+      allowStale: true,
+      fetchMethod: async (address: string) => {
+        return await baseService.getAccount(address)
+      },
+    })
+  }
+
+  getAccount = (address: string): Promise<string> => {
+    return traceAsyncFunction('CachingAccountService - getAccount', async () => {
+      const dek = await this.cache.fetch(address)
+
+      if (dek === undefined) {
+        // TODO decide which error ot use here
+        throw new OdisError(ErrorMessage.FAILURE_TO_GET_DEK)
+      }
+      return dek
+    })
+  }
+}
+
+// tslint:disable-next-line:max-classes-per-file
+export class ContractKitAccountService implements AccountService {
```

Comment:
I made some changes to this part of the code in the Signer that would be worth taking a look at